### PR TITLE
Allow one signature per person

### PR DESCRIPTION
**Pull request by andyg0808**
Let's restrict the number of signatures each person can make to one. This will be slightly more bother when carrying over a CR, but it will reduce the risk of accidentally deploying a pull which only one person has signed off on.

We encountered an incident where a person signed off twice, once before a commit was pushed and once after, but because the commit date was before the first signoff date, both signoffs were considered valid. The problem with a signoff not being invalidated by a commit which is pushed after the signoff is still not solved, but this change at least reduces the risk of such an incident resulting in insufficiently-reviewed code being deployed.

### CODE REVIEW

File: public/js/Pull.js [hunk 1]
```diff
@@ -24,15 +24,12 @@ _.extend(Pull.prototype, {
          groups.user = null;
 
          signatures.forEach(function(signature) {
-            if (signature.data.active) {
-               groups.current.push(signature);
-               users[signature.data.user.id] = true;
-
-               if (utils.mySig(signature)) {
-                  groups.user = signature;
+            if (!users[signature.data.user.id]) {
```
Comment (jarstelfox) [inline public/js/Pull.js:27]: I would invert this and early return to reduce nesting.

File: public/js/Pull.js [hunk 2]
```diff
@@ -24,15 +24,12 @@ _.extend(Pull.prototype, {
...
+               if (signature.data.active) {
+                  groups.current.push(signature);
+               } else {
+                  groups.old.push(signature);
```
Comment (jarstelfox) [inline public/js/Pull.js:31]: Reduce nesting:
```suggestion
               const groups = signature.data.active ? groups.current : groups.old;
               groups.push(signature);
```

Comment (jarstelfox) [inline public/js/Pull.js:31]: Reduce nesting:
```suggestion
               const groups = signature.data.active ? groups.current : groups.old;
               groups.push(signature);
```

FYI: I can not include the last brace in my suggestion so do this by hand please

**Comment by djmetzle** [comment]
CR :gear: though @jarstelfox has some good cleanup ideas.

File: public/js/Pull.js (hunk 2 above)
Comment (davidrans) [reply public/js/Pull.js:31]: This is fewer lines, but not more readable IMO :shrug:

Comment (jarstelfox) [reply public/js/Pull.js:31]: IMO it's more readable. Namely, you don't have to keep the state in your head. Every time I see an if, I mentally note that we are forking the state and need to consider other cases. Between inversion and ternaries, it becomes more clear to me that we are handling + edge case and picking a bucket.

That being said it's up to Andrew here

**Comment by sctice** [comment]
I'm finding this logic a bit difficult to follow, so I may very well be mistaken, but it looks to me like the new version could fail to recognize an active signature for someone who also has an inactive signature that was processed first. It looks to me like whichever case we encounter first, inactive or active, that's the end of the story for that user. For the same reasons, I think we're changing how the `groups.user` signature gets assigned.

I had a hard time when just looking at the diff, so in case it helps anyone else, here's the previous version in full:

https://github.com/iFixit/pulldasher/blob/236d7bf246be021570268ea83e38967637071a38/public/js/Pull.js#L14-L45

CR 🎏 
dev_block ✋ on what looks to me like an unintended change in behavior.

**Comment by andyg0808** [comment]
dev_block :+1: 
on comments. Good points!

**Comment by andyg0808** [comment]
@sctice I believe the signatures are handled ordered by date descending. I have two reasons for this belief:
1. If they were not, the existing code would potentially add an "expired" sign off and then a "current" signoff for the same user, if they were encountered in that order.
2. The code to fetch signatures from the DB sorts them in descending date order:
https://github.com/iFixit/pulldasher/blob/e040c3b2b741d3af11b3ec7d8cfe91a5c4a3036a/lib/db-manager.js#L345-L356
Although I can't guarantee that that's how the data is fetched on every code path, I seem to remember that the DB is integral to how the backend works, so I'm guessing that method may, in fact, be the only way signature data gets to the frontend.

As far as I've seen, the pulls are refreshed as objects, so I don't think there's any update of the signature list item-by-item; I think we're sending the entire new pull to the frontend every time a signature is updated.

File: public/js/Pull.js (hunk 1 above)
Comment (andyg0808) [reply public/js/Pull.js]: Good idea!
3a87c82

File: public/js/Pull.js (hunk 2 above)
Comment (andyg0808) [reply public/js/Pull.js]: I'm fine with this the way it is.

**Comment by andyg0808** [comment]
un_dev_block :+1:

**Comment by djmetzle** [comment]
CR :pen:

**Comment by jyee27** [comment]
CR :coffee:

**Pull request merged**